src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs

```diff
@@ -489,58 +504,68 @@ private void ProcessPendingSubmissions()
+                        return;
                     }
-                }));
-            }
 
-            public Task SubmitAsync()
-            {
-                RequiresLanguageBuffer();
+                    // get command to save to history before calling FinishCurrentSubmissionInput
+                    // as it adds newline at the end
+                    var historySpan = _window._currentLanguageBuffer.CurrentSnapshot.GetExtent();
+                    FinishCurrentSubmissionInput();
 
-                // TODO: queue submission
-                // Ensure that the REPL doesn't try to execute if it is already
-                // executing.  If this invariant can no longer be maintained more of
-                // the code in this method will need to be bullet-proofed
-                if (State == State.ExecutingInput)
-                {
-                    return Task.FromResult<object>(null);
-                }
+                    _window._history.UncommittedInput = null;
 
-                // get command to save to history before calling FinishCurrentSubmissionInput
-                // as it adds newline at the end
-                var historySpan = _window._currentLanguageBuffer.CurrentSnapshot.GetExtent();
-                FinishCurrentSubmissionInput();
+                    var snapshotSpan = _window._currentLanguageBuffer.CurrentSnapshot.GetExtent();
+                    var trimmedSpan = snapshotSpan.TrimEnd();
 
-                _window._history.UncommittedInput = null;
+                    if (trimmedSpan.Length == 0)
+                    {
+                        // TODO: reuse the current language buffer
+                        PrepareForInput();
+                        return;
+                    }
+                    else
+                    {
+                        _window._history.Add(historySpan);
+                        State = State.ExecutingInput;
 
-                var snapshotSpan = _window._currentLanguageBuffer.CurrentSnapshot.GetExtent();
-                var trimmedSpan = snapshotSpan.TrimEnd();
+                        StartCursorTimer();
 
-                if (trimmedSpan.Length == 0)
-                {
-                    // TODO: reuse the current language buffer
-                    PrepareForInput();
-                    return Task.FromResult<object>(null);
-                }
-                else
-                {
-                    _window._history.Add(historySpan);
-                    State = State.ExecutingInput;
+                        var executionResult = await _window._evaluator.ExecuteCodeAsync(snapshotSpan.GetText()).ConfigureAwait(true);
+                        Debug.Assert(_window.OnUIThread()); // ConfigureAwait should bring us back to the UI thread.
 
-                    StartCursorTimer();
+                        Debug.Assert(State == State.ExecutingInput || State == State.Resetting, $"Unexpected state {State}");
```

Comment:
We should be able to make a stronger guarantee here since the submissions and executions are all initiated on the UI thread. Specifically, a reset should be treated as a submission, and submissions should be compiled and executed sequentially.